rama-http/src/service/fs/serve_dir/open_file.rs

```diff
@@ -41,14 +45,122 @@ pub(super) struct FileOpened {
 pub(super) enum FileRequestExtent {
     Full(File, Metadata),
     Head(Metadata),
+    Embedded(Vec<u8>),
+    EmbeddedHead(u64), // Content length for HEAD requests on embedded files
+}
+
+impl FileRequestExtent {
+    pub(super) fn reader(self) -> Option<Box<dyn AsyncRead + Send + Sync + Unpin>> {
+        match self {
+            Self::Head(_) | Self::EmbeddedHead(_) => None,
+            Self::Full(file, _) => Some(Box::new(file)),
+            Self::Embedded(content) => Some(Box::new(Cursor::new(content))),
+        }
+    }
+
+    pub(super) fn range_reader_with_offset(
+        self,
+        offset: u64,
+        range_size: u64,
+    ) -> Option<Box<dyn AsyncRead + Send + Sync + Unpin>> {
+        match self {
+            Self::Head(_) | Self::EmbeddedHead(_) => None,
+            Self::Full(file, _) => Some(Box::new(file.take(range_size))), // File is already seeked
+            Self::Embedded(content) => {
+                let start = offset as usize;
+                let end = (offset + range_size) as usize;
+                let slice = if start < content.len() {
+                    &content[start..end.min(content.len())]
+                } else {
+                    &[]
+                };
+                Some(Box::new(Cursor::new(slice.to_vec())))
+            }
+        }
+    }
+
+    pub(super) fn get_size(&self) -> u64 {
+        match self {
+            Self::Head(meta) | Self::Full(_, meta) => meta.len(),
+            Self::Embedded(content) => content.len() as u64,
+            Self::EmbeddedHead(size) => *size,
+        }
+    }
+}
+
+pub(super) fn open_file_embedded(
+    base: &Dir,
+    mut path_to_file: PathBuf,
+    req: &Request,
+    negotiated_encodings: &[QualityValue<Encoding>],
+    range_header: Option<&str>,
+    buf_chunk_size: usize,
+) -> io::Result<OpenFileOutput> {
+    // Check if this is a directory request (empty path or directory path)
+    // If so, try to serve index.html
+    if path_to_file.as_os_str().is_empty() || path_to_file.to_str() == Some(".") {
+        path_to_file = PathBuf::from("index.html");
+    }
+
+    let mime = mime_guess::from_path(&path_to_file)
+        .first_raw()
+        .map(HeaderValue::from_static)
+        .unwrap_or_else(|| HeaderValue::from_str(mime::APPLICATION_OCTET_STREAM.as_ref()).unwrap());
+
+    let maybe_encoding = preferred_encoding(&mut path_to_file, negotiated_encodings);
+
+    let Some(file) = base.get_file(&path_to_file) else {
+        return Ok(OpenFileOutput::FileNotFound);
+    };
+
+    let last_modified = file
```

Comment:
there is a lot of overlap between file embedded and regular file. Not only here but in general. Do try to keep a check on the duplication. There are are probably some more clever abstractions we can introduce to limit that kind of code duplication